src/snowflake/snowpark/mock/_snowflake_data_type.py

```diff
@@ -234,6 +242,74 @@ def calculate_type(c1: ColumnType, c2: Optional[ColumnType], op: Union[str]):
     )
 
 
+def coerce_t1_into_t2(t1: DataType, t2: DataType) -> Optional[DataType]:
```

Comment:
This function looks mostly correct to me. I suspect we may find additional rules in the future. Can you leave a comment on how you found and verified that these rules were true?